baybe/recommenders/pure/bayesian/botorch.py

```diff
@@ -151,31 +168,176 @@ def _recommend_continuous(
+                f"This method expects a subspace object with constraints of type "
+                f"{ContinuousCardinalityConstraint.__name__}. For a subspace object "
+                f"without constraints of type"
+                f" {ContinuousCardinalityConstraint.__name__}, "
+                f"{self._recommend_continuous_without_cardinality_constraints.__name__}."  # noqa
+            )
+
+        acqf_values_all: list[Tensor] = []
+        points_all: list[Tensor] = []
+
+        def append_recommendation_for_inactive_parameters_setting(
+            inactive_parameters: tuple[str, ...],
+        ):
+            """Append the recommendation for each inactive parameter configuration.
+
+            Args:
+                inactive_parameters: A list of inactive parameters.
+            """
+            # Create a new subspace by ensuring all active parameters being
+            # non-zeros.
+            subspace_continuous_with_active_params = (
+                subspace_continuous._ensure_nonzero_parameters(inactive_parameters)
+            )
+            # Optimize the acquisition function
+            (
+                points_i,
+                acqf_values_i,
+            ) = self._recommend_continuous_without_cardinality_constraints(
+                subspace_continuous_with_active_params,
+                batch_size,
+                inactive_parameters,
+            )
+            # Append recommendation list and acquisition function values
+            points_all.append(points_i.unsqueeze(0))
+            acqf_values_all.append(acqf_values_i.unsqueeze(0))
+
+        # Below we start recommendation
+        if (
+            subspace_continuous.n_inactive_parameter_combinations
+            > self.n_threshold_inactive_parameters_generator
+        ):
+            # When the combinatorial list is too large, randomly set some parameters
+            # inactive.
+            for _ in range(self.n_threshold_inactive_parameters_generator):
+                inactive_params_sample = tuple(
+                    subspace_continuous._sample_inactive_parameters(1)[0]
+                )
+                append_recommendation_for_inactive_parameters_setting(
+                    inactive_params_sample
+                )
+        else:
+            # When the combinatorial list is not too large, iterate the combinatorial
+            # list of all possible inactive parameters.
+            for (
+                inactive_params_generator
+            ) in subspace_continuous.inactive_parameter_combinations():
+                append_recommendation_for_inactive_parameters_setting(
+                    inactive_params_generator
+                )
+
+        # Find the best option
+        points = torch.cat(points_all)[torch.argmax(torch.cat(acqf_values_all)), :]
+        acqf_values = torch.max(torch.cat(acqf_values_all))
+        return points, acqf_values
+
+    def _recommend_continuous_without_cardinality_constraints(
```

Comment:
I find the design here a bit weird. The method is called `..._without_cardinality_constraint` but somehow still takes care of inactive parameters internally. My general question here is: instead of fixing inactive values to 0.0, can't we apply the same logic of creating a reduced space instead? If that was possible, I think it would also drastically reduce the amount necessary code

Reply:
I believe design in the revised version is not that misleading. Regarding the inactive parameters, please see my reply above https://github.com/emdgroup/baybe/pull/301#discussion_r1730886062

Reply:
naively I also expected that the `without` function should basically be our old function content-wise (just applied to a reduced/modified subspace) -> can you explain why the meddling with fixed parameters is necessary below? Is it due to numerical things?

Reply:
I've refactored the entire PR, the logic is much clearer now. Will close here and ping you once it's ready for review.